File: solutions/0401-0500/0412-fizz-buzz/solution.cpp
#include "solution.hpp"

namespace leetcode::p0412 {

std::vector<std::string> solve(int n) {
    std::vector<std::string> result;
    result.reserve(static_cast<std::size_t>(n));

    for (int i = 1; i <= n; ++i) {
        if (i % 15 == 0) {
            result.emplace_back("FizzBuzz");
        } else if (i % 3 == 0) {
            result.emplace_back("Fizz");
        } else if (i % 5 == 0) {
            result.emplace_back("Buzz");
        } else {
            result.push_back(std::to_string(i));
        }
    }

    return result;
}

} // namespace leetcode::p0412

File: solutions/0401-0500/0412-fizz-buzz/solution.ml
type error = Out_of_range of int

let max_n = 10_000

let fizz_buzz_word n =
  match (n mod 3, n mod 5) with
  | 0, 0 -> "FizzBuzz"
  | 0, _ -> "Fizz"
  | _, 0 -> "Buzz"
  | _, _ -> string_of_int n

let solve n =
  if n < 1 || n > max_n then Error (Out_of_range n)
  else Ok (List.init n (fun i -> fizz_buzz_word (i + 1)))
